File: Java/src/part1/Backtracking.java
package part1;

import java.util.HashSet;
import java.util.Scanner;

import part1.Board;

public class Backtracking {

	static int n;
	static int counter = 0;
	static long start;
	static long end;
	
	public static void main(String[] args) {
		
		Scanner user_input = new Scanner(System.in);
		
		String[] starting_positions_as_strings = user_input.nextLine().split(" ");
		
		user_input.close();
		
		int[] starting_positions = new int[starting_positions_as_strings.length];
		
		
		for(int i = 0; i < starting_positions_as_strings.length; i++) {
			starting_positions[i] = Integer.parseInt(starting_positions_as_strings[i]) - 1;
		}
		
		n = starting_positions.length;
		
		
		Board startingBoard = new Board(starting_positions);
		
		if(StartingPositionValidator.starting_position_valid(starting_positions)) {
			System.out.println("Valid starting position. Starting backtracking search for feasable solutions");
			start = System.currentTimeMillis();
			try_col(startingBoard);
			end = System.currentTimeMillis();
			System.out.print("Total time used: ");
			System.out.println((end - start) / 1000F);
			System.out.print("Number of solutions found: ");
			System.out.println(counter);
		} else {
			System.out.println("Invalid starting position.");
		}
		
	}
	
	
	private static void try_col(Board board) {
		
		int col = board.getCol();
		int[] colList = board.getColList();
		HashSet<Integer> rowsSet = board.getRowsSet();
		
		if (col == n) {
			counter++;
			System.out.println(board.toString());
		}
		
		if (rowsSet.size() == 0) {
			return;
		}

		for(int row : rowsSet) {
			if(!DiagonalChecker.diagonal_conflict(row, col, colList)) {
				int[] newColList = colList;
				newColList[col] = row;
				HashSet<Integer> newRowsSet = new HashSet<Integer>(rowsSet);
				newRowsSet.remove(row);
				
				try_col(new Board(col + 1, newColList, newRowsSet));
			}
		}
	}
}

File: Python/backtracking.py
from time import time
from input_handler import starting_positions_backtracking_algorithm, get_N_from_user
from output_handler import show_solutions
from utils import no_diagonal_conflict, get_all_numbers_not_used, subtract_one_from_list, starting_positions_valid

N = get_N_from_user()
SOLUTIONS = []
STEP_BY_STEP = []

def backtracking(col, col_list, rows_set):

    (len(SOLUTIONS) == 0) and STEP_BY_STEP.append(col_list[:])

    if col == N:
        SOLUTIONS.append(col_list[:])
        return
 
    for row in rows_set:
        if no_diagonal_conflict(row, col, col_list):
            col_list.append(row)
            rows_set.remove(row)
            backtracking(col + 1, col_list, rows_set)
            rows_set.add(row)
            del col_list[-1]
    return
 
def main(starting_positions):
    rows_set = get_all_numbers_not_used(starting_positions, N)
    if starting_positions_valid(starting_positions, N):
        print "Valid starting positions, starting backtracking search for feasable solution"
        backtracking(len(starting_positions), starting_positions, rows_set)

starting_positions = subtract_one_from_list(starting_positions_backtracking_algorithm(N))
start = time()
main(starting_positions)
show_solutions(STEP_BY_STEP, SOLUTIONS, time() - start, N, True)
